packages/editor/src/components/post-actions/set-as-homepage.js

```diff
@@ -0,0 +1,216 @@
+				'postType',
+				'page',
+				siteSettings?.page_on_front
+			);
+			return {
+				showOnFront: siteSettings?.show_on_front,
+				currentHomePage: currentHomePageItem,
+				isSavingSiteSettings: isSavingEntityRecord( 'root', 'site' ),
+			};
+		}
+	);
+	const isPageDraftRef = useRef( item.status === 'draft' );
+	const isPageDraft = isPageDraftRef.current;
+	const currentHomePageTitle = currentHomePage
+		? getItemTitle( currentHomePage )
+		: '';
+
+	const { saveEditedEntityRecord, saveEntityRecord } =
+		useDispatch( coreStore );
+	const { createSuccessNotice, createErrorNotice } =
+		useDispatch( noticesStore );
+
+	async function onSetPageAsHomepage( event ) {
+		event.preventDefault();
+
+		try {
+			// If selected page is set to draft, publish the page.
+			if ( isPageDraft ) {
+				await saveEntityRecord( 'postType', 'page', {
+					...item,
+					status: 'publish',
+				} );
+			}
+
+			// Save new home page settings.
+			await saveEntityRecord( 'root', 'site', {
+				page_on_front: item.id,
+				show_on_front: 'page',
+			} );
+
+			closeModal?.();
+
+			await saveEditedEntityRecord( 'root', 'site', undefined, {
+				page_on_front: item.id,
+				show_on_front: 'page',
+			} );
+
+			createSuccessNotice( __( 'Homepage updated' ), {
+				type: 'snackbar',
+			} );
+		} catch ( error ) {
+			const typedError = error;
+			const errorMessage =
+				typedError.message && typedError.code !== 'unknown_error'
+					? typedError.message
+					: __( 'An error occurred while setting the homepage' );
+			createErrorNotice( errorMessage, { type: 'snackbar' } );
+		}
+	}
+
+	const renderModalBody = () => {
+		if ( 'posts' === showOnFront ) {
+			return (
+				<>
+					<Text>
+						{ sprintf(
+							// translators: %s: title of the page to be set as the homepage.
+							__(
+								'Set "%s" as the site homepage? This will replace the current homepage which is set to display latest posts.'
+							),
+							pageTitle
+						) }
+					</Text>
+				</>
+			);
+		}
+
+		if ( isPageDraft ) {
+			return (
+				<>
+					<Text>
+						{ sprintf(
+							// translators: %s: title of the page to be set as the homepage.
+							__(
+								'"%s" is a draft and will be published automatically if set as the homepage. Publish page and set as the site homepage?'
+							),
+							pageTitle
+						) }
+					</Text>
+				</>
+			);
+		}
+
+		const modalTranslatedString =
+			// translators: %1$s: title of page to be set as the home page. %2$s: title of the current home page.
+			__(
+				'Set "%1$s" as the site homepage? This will replace the current homepage: "%2$s"'
+			);
+
+		return (
+			<Text>
+				{ sprintf(
+					modalTranslatedString,
+					pageTitle,
+					currentHomePageTitle
+				) }{ ' ' }
+			</Text>
+		);
+	};
+
+	// translators: Button label to confirm setting the specified page as the homepage.
+	let modalButtonLabel = __( 'Set homepage' );
+	if ( isPageDraft ) {
+		modalButtonLabel = __( 'Publish and set homepage' );
+	}
```

Comment:
```suggestion
	const modalButtonLabel = isPageDraft ? __( 'Publish and set homepage' ) : __( 'Set homepage' );
```

Nit: let's avoid the temp var.

Reply:
Unless you were doing this to provide better context for translators. In which case it's fine as is.

Reply:
Done in https://github.com/WordPress/gutenberg/pull/65426/commits/b1707e45a2fd33be67f957b43dda63e79b872ae4.